File: Digital Curriculum/src/app/lib/betaChecklist.ts
import {
  collection,
  doc,
  getDoc,
  getDocs,
  limit,
  query,
  serverTimestamp,
  setDoc,
  where,
} from "firebase/firestore";
import { db } from "./firebase";

/**
 * The external beta tester checklist for the Digital Curriculum round.
 *
 * Every step is marked off by evidence the tester actually did the thing —
 * there is no manual tick. Where a step already leaves a trace in Firestore we
 * read that trace; the two that do not (looking around the dashboard, filing a
 * report) are stamped onto the tester's own user doc at the moment they happen,
 * because `beta_feedback` is staff-read-only and a tester cannot query their
 * own reports back.
 */

export type BetaStepId =
  | "account"
  | "onboarding"
  | "dashboard"
  | "start"
  | "lesson"
  | "application"
  | "dataroom"
  | "event"
  | "dm"
  | "report"
  | "app";

export interface BetaStep {
  id: BetaStepId;
  /** Short label for the collapsed rail. */
  title: string;
  detail: string[];
  note?: { label: string; text: string };
}

/** Field on `users/{uid}` holding stamps for steps with no other trace. */
export const BETA_CHECKLIST_FIELD = "beta_checklist";

/** Seconds on the dashboard before "look around the dashboard" counts as done. */
export const DASHBOARD_DWELL_SECONDS = 30;

export const BETA_STEPS: BetaStep[] = [
  {
    id: "account",
    title: "Create your account",
    detail: [
      "Open the beta link we emailed you.",
      "Choose “Don’t have an account? Sign up”, then enter your email and a password.",
    ],
  },
  {
    id: "onboarding",
    title: "Work through onboarding",
    detail: [
      "We are deliberately not giving instructions — we want to see whether it explains itself.",
      "Anywhere you hesitate or have to guess, tell us.",
      "Check your inbox for the “Welcome to MORTAR Masters Online” email.",
    ],
  },
  {
    id: "dashboard",
    title: "Look around the dashboard",
    detail: [
      "Read it the way you would on a normal day.",
      "Anything unclear, mislabelled or missing, report it with the bug icon.",
    ],
    note: {
      label: "Ticks itself",
      text: "This one completes once you have spent a little time here.",
    },
  },
  {
    id: "start",
    title: "Start the Beta Testing course",
    detail: [
      "Open Curriculum and find the Beta Testing course.",
      "Open it and click Start Lesson — the first slide is enough to begin.",
    ],
    note: {
      label: "Already yours",
      text: "The course is free and already assigned to your account, so there is nothing to buy. This step ticks as soon as you open the first lesson.",
    },
  },
  {
    id: "lesson",
    title: "Take the lesson and both quizzes",
    detail: [
      "Click Start Lesson and read through it.",
      "Start the quiz, pick your favourite celebrity, and submit.",
      "Choose “Analyze my response with AI”, read the feedback, then Finish.",
      "Take the course quiz as well, then close it.",
    ],
  },
  {
    id: "application",
    title: "Submit your alumni application",
    detail: [
      "In Curriculum you will now see Alumni Application — click Apply Now.",
      "Add any availability slots you like. This is test data.",
    ],
    note: {
      label: "We reply later",
      text: "A real person reviews this. Expect two emails within one working day — keep the second, it carries the access code you need at step 11.",
    },
  },
  {
    id: "dataroom",
    title: "Collect your survey and certificate",
    detail: [
      "Open Data Room → Corporate Documents, find your Beta Testing Survey, and download it.",
      "Go to Skills & Certificates and check your new skill is listed.",
      "Preview the certificate, then download it.",
      "Click Add to LinkedIn and read the instructions — you do not have to post it.",
    ],
  },
  {
    id: "event",
    title: "Register for an event",
    detail: [
      "Go to the Community Hub, and in Events choose View All Events.",
      "Pick one, click View & Register, read the details, and register.",
    ],
  },
  {
    id: "dm",
    title: "Send us a direct message",
    detail: [
      "From the Community Hub, use DM MORTAR to send us anything at all.",
      "Come back later and check our reply arrived and reads properly.",
    ],
    note: {
      label: "We reply later",
      text: "Do not wait on this one — carry on. We answer within one working day.",
    },
  },
  {
    id: "report",
    title: "Report anything outstanding",
    detail: [
      "Go back to anything that confused you and report it while it is fresh.",
      "Tell us the one thing you would change first if it were up to you.",
    ],
    note: {
      label: "Bug icon",
      text: "Bottom-right of any screen. This step completes when your first report is filed.",
    },
  },
  {
    id: "app",
    title: "Install the app and sign in",
    detail: [
      "Open the email titled “Your MORTARverse App Access Code Has Arrived” and copy the code.",
      "Install the app using the download link in that email.",
      "Sign in with the same email you used here, plus your access code.",
      "Stop once you reach the Mortarverse screen — round two picks up from there.",
    ],
    note: {
      label: "Needs step 6",
      text: "This waits on your admission email. If the code is refused it has expired — reply to that email for a fresh one rather than making a second account.",
    },
  },
];

export type BetaStepState = Record<BetaStepId, boolean>;

function emptyState(): BetaStepState {
  return BETA_STEPS.reduce((acc, step) => {
    acc[step.id] = false;
    return acc;
  }, {} as BetaStepState);
}

/** True when this account has been flagged as an external beta tester. */
export function isBetaTester(userData: Record<string, unknown> | undefined): boolean {
  return userData?.beta_tester === true;
}

async function anyDocExists(
  collectionName: string,
  field: string,
  value: string
): Promise<boolean> {
  try {
    const snap = await getDocs(
      query(collection(db, collectionName), where(field, "==", value), limit(1))
    );
    return !snap.empty;
  } catch (e) {
    // A denied or offline read must not blank the whole checklist — treat the
    // single step as "not yet" and let the others resolve.
    console.warn(`[betaChecklist] ${collectionName} lookup failed`, e);
    return false;
  }
}

async function anySubcollectionDoc(uid: string, sub: string): Promise<boolean> {
  try {
    const snap = await getDocs(query(collection(db, "users", uid, sub), limit(1)));
    return !snap.empty;
  } catch (e) {
    console.warn(`[betaChecklist] users/${uid}/${sub} lookup failed`, e);
    return false;
  }
}

async function hasRegisteredForAnyEvent(uid: string): Promise<boolean> {
  const inCollection = async (name: string) => {
    try {
      const snap = await getDocs(
        query(collection(db, name), where("registered_users", "array-contains", uid), limit(1))
      );
      return !snap.empty;
    } catch (e) {
      console.warn(`[betaChecklist] ${name} registration lookup failed`, e);
      return false;
    }
  };
  const [curriculum, mobile] = await Promise.all([
    inCollection("events"),
    inCollection("events_mobile"),
  ]);
  return curriculum || mobile;
}

interface CourseProgressSummary {
  /** A progress doc exists at all — see {@link courseProgressSummary}. */
  started: boolean;
  /** At least one course reads as finished. */
  completed: boolean;
}

/**
 * Started and finished state for the tester's courses, in one query.
 *
 * `courseProgress/{uid}_{courseId}` is created lazily by
 * `initializeCourseProgress` the first time a learner views a slide, so the
 * document's mere existence is the "started the course" signal — there is no
 * separate enrolment record to look for, and the beta course is free and
 * pre-assigned, so nothing is purchased along the way.
 *
 * Filtered on `userId` alone and finished off in memory: adding `completed` to
 * the query would need a composite index, and a tester has only a handful of
 * progress docs. `completed` is the authoritative flag, but a course sitting at
 * 100 without it counts too — that gap is exactly what a beta finds.
 *
 * Assumes the beta cohort is assigned a single course, so "any progress doc"
 * means the beta course. Scope this to a course id if that stops being true.
 */
async function courseProgressSummary(uid: string): Promise<CourseProgressSummary> {
  try {
    const snap = await getDocs(
      query(collection(db, "courseProgress"), where("userId", "==", uid))
    );
    if (snap.empty) return { started: false, completed: false };
    const completed = snap.docs.some((d) => {
      const data = d.data() as { completed?: unknown; progress?: unknown };
      if (data.completed === true) return true;
      return typeof data.progress === "number" && data.progress >= 100;
    });
    return { started: true, completed };
  } catch (e) {
    console.warn("[betaChecklist] courseProgress lookup failed", e);
    return { started: false, completed: false };
  }
}

/**
 * Read every completion signal for one tester.
 *
 * `userData` is the already-loaded `users/{uid}` document where the caller has
 * one — the dashboard does — so this adds four queries rather than five reads.
 */
export async function evaluateBetaChecklist(
  uid: string,
  userData?: Record<string, unknown>
): Promise<BetaStepState> {
  const state = emptyState();
  if (!uid) return state;

  let data = userData;
  if (!data) {
    try {
      const snap = await getDoc(doc(db, "users", uid));
      data = snap.data();
    } catch (e) {
      console.warn("[betaChecklist] user doc read failed", e);
      return state;
    }
  }

  const stamps = (data?.[BETA_CHECKLIST_FIELD] ?? {}) as Record<string, unknown>;

  const [hasApplication, hasSurvey, hasCertificate, hasEvent, hasDm, course] =
    await Promise.all([
      anyDocExists("GraduationApplications", "userId", uid),
      anySubcollectionDoc(uid, "surveyResponses"),
      anySubcollectionDoc(uid, "certificates"),
      hasRegisteredForAnyEvent(uid),
      anyDocExists("Digital Student DMs", "uid", uid),
      courseProgressSummary(uid),
    ]);

  // Reaching the dashboard at all means the account exists and is usable.
  state.account = true;
  state.onboarding = data?.onboarding_status === "complete";
  state.dashboard = Boolean(stamps.dashboard_viewed_at);
  state.start = course.started;
  state.lesson = course.completed;
  state.application = hasApplication;
  state.dataroom = hasSurvey && hasCertificate;
  state.event = hasEvent;
  state.dm = hasDm;
  state.report = Boolean(stamps.reported_at);
  state.app = data?.expansion_mobile_app_account_created === true;

  return state;
}

/**
 * Record a step that leaves no other trace. Best-effort: a tester whose write
 * is rejected still gets a working checklist, just without that tick.
 */
export async function stampBetaChecklist(
  uid: string,
  key: "dashboard_viewed_at" | "reported_at"
): Promise<void> {
  if (!uid) return;
  try {
    await setDoc(
      doc(db, "users", uid),
      { [BETA_CHECKLIST_FIELD]: { [key]: serverTimestamp() } },
      { merge: true }
    );
  } catch (e) {
    console.warn(`[betaChecklist] could not stamp ${key}`, e);
  }
}

File: ExpansionNetworkApp/expansion_network/lib/beta/beta_checklist.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

/// The external beta tester checklist for the MORTARverse round.
///
/// Nothing here is tickable by hand. Each step reads the trace the tester's own
/// actions already left in Firestore, so a step crosses itself off the moment
/// the work is genuinely done — the same contract as the Digital Curriculum
/// rail on the web.
///
/// Every read below is a single-field query the member is already allowed to
/// make under `firestore.rules`, so none of this needs a rules change or a
/// composite index.
class BetaStep {
  const BetaStep({
    required this.id,
    required this.title,
    required this.detail,
    this.note,
    this.optional = false,
  });

  final String id;

  /// Short label for the collapsed banner.
  final String title;

  final List<String> detail;

  /// Extra warning or expectation, shown under the detail when expanded.
  final String? note;

  /// Optional steps never block "all done" and are counted separately.
  final bool optional;
}

const List<BetaStep> kBetaSteps = [
  BetaStep(
    id: 'profile',
    title: 'Sweep the screen and finish your profile',
    detail: [
      'Swipe the whole top carousel and check each screen opens.',
      'Open Complete Profile, fill in your bio properly, and save.',
      'Check your badge shows under your profile, then open Badges.',
    ],
    note: 'Crosses off once your profile reaches 100%.',
  ),
  BetaStep(
    id: 'event',
    title: 'Register for the test event',
    detail: [
      'Scroll to the lowest widget and tap Register.',
      'Check the calendar pops up and the details look right.',
      'Look at the RSVP list and try messaging someone on it.',
    ],
  ),
  BetaStep(
    id: 'matching',
    title: 'Run smart matching',
    detail: [
      'Enter the Networking Hall and tap Run Smart Matching, then Start Matching.',
      'Message one of your matches and open their profile.',
      'Check Mortar Shop opens, and that the Mortar Info video plays with sound.',
    ],
  ),
  BetaStep(
    id: 'explore',
    title: 'Post a job and a skill',
    detail: [
      'Open Explore and check the All, Jobs and Skills tabs each fill up.',
      'Search for someone, open their profile, and message them.',
      'Tap Add to post a job, then again to post a skill.',
    ],
    note: 'Needs both a job and a skill before it crosses off.',
  ),
  BetaStep(
    id: 'feed',
    title: 'Post to the home feed',
    detail: [
      'Open Recent Activity and tap View All.',
      'Like a few posts and reply to one.',
      'Tap Add, write a post, attach an image, and publish.',
    ],
  ),
  BetaStep(
    id: 'community',
    title: 'Join a community and start one',
    detail: [
      'Under Your Communities, tap Groups and join any group.',
      'Publish a message, then comment on your own message.',
      'Tap Add and create a community of your own.',
      'In Events, register for one and check the Registered tab.',
    ],
    note: 'Needs both joining and creating before it crosses off.',
  ),
  BetaStep(
    id: 'conference',
    title: 'Buy a Conference Center ticket',
    detail: [
      'Go to the Conference Center and buy a ticket to Our North Star.',
      'Watch for the confirmation email with your access code.',
    ],
    note: 'Real money — your card is charged. Skip it if you would rather not; '
        'nothing else depends on it.',
    optional: true,
  ),
];

/// Which steps are done, and the counts the banner shows.
class BetaChecklistState {
  const BetaChecklistState(this.done);

  const BetaChecklistState.empty() : done = const {};

  final Map<String, bool> done;

  bool isDone(String id) => done[id] ?? false;

  static Iterable<BetaStep> get _required =>
      kBetaSteps.where((s) => !s.optional);

  int get totalRequired => _required.length;

  int get completedRequired =>
      _required.where((s) => isDone(s.id)).length;

  bool get allRequiredDone => completedRequired == totalRequired;

  double get progress =>
      totalRequired == 0 ? 0 : completedRequired / totalRequired;
}

/// Reads every completion signal for the signed-in tester.
class BetaChecklistService {
  BetaChecklistService({FirebaseFirestore? firestore, FirebaseAuth? auth})
      : _db = firestore ?? FirebaseFirestore.instance,
        _auth = auth ?? FirebaseAuth.instance;

  final FirebaseFirestore _db;
  final FirebaseAuth _auth;

  /// True when at least one document matches. A denied or offline read resolves
  /// to false rather than throwing, so one unavailable signal cannot blank the
  /// whole checklist.
  Future<bool> _any(Query<Map<String, dynamic>> query) async {
    try {
      final snap = await query.limit(1).get();
      return snap.docs.isNotEmpty;
    } catch (e) {
      return false;
    }
  }

  /// [profileComplete] comes from `MortarverseSignals`, which the chooser
  /// already streams — no point reading the profile twice.
  ///
  /// [openConferenceId] is the conference currently open, if any. With no
  /// conference open there is no ticket to buy, so that step stays unticked.
  Future<BetaChecklistState> load({
    required bool profileComplete,
    String? openConferenceId,
  }) async {
    final uid = _auth.currentUser?.uid;
    if (uid == null) return const BetaChecklistState.empty();

    final results = await Future.wait([
      _any(_db.collection('events_mobile').where(
            'registered_users',
            arrayContains: uid,
          )),
      _any(_db.collection('users').doc(uid).collection('expansion_matches')),
      _any(_db.collection('expansion_jobs').where('author_id', isEqualTo: uid)),
      _any(_db.collection('expansion_skills').where('author_id', isEqualTo: uid)),
      _any(_db.collection('feed_posts').where('author_id', isEqualTo: uid)),
      _any(_db.collection('groups_mobile').where(
            'GroupMembers',
            arrayContains: uid,
          )),
      _any(_db.collection('groups_mobile').where('createdBy', isEqualTo: uid)),
      _hasConferenceTicket(uid, openConferenceId),
    ]);

    final registeredForEvent = results[0];
    final hasMatches = results[1];
    final postedJob = results[2];
    final postedSkill = results[3];
    final postedToFeed = results[4];
    final joinedGroup = results[5];
    final createdGroup = results[6];
    final hasTicket = results[7];

    return BetaChecklistState({
      'profile': profileComplete,
      'event': registeredForEvent,
      'matching': hasMatches,
      // The step asks for both, so both are required before it crosses off.
      'explore': postedJob && postedSkill,
      'feed': postedToFeed,
      'community': joinedGroup && createdGroup,
      'conference': hasTicket,
    });
  }

  Future<bool> _hasConferenceTicket(String uid, String? conferenceId) async {
    if (conferenceId == null) return false;
    try {
      final snap = await _db
          .collection('conferences')
          .doc(conferenceId)
          .collection('attendees')
          .doc(uid)
          .get();
      return snap.exists;
    } catch (e) {
      return false;
    }
  }
}
